### PR TITLE
feat(native-filters): Filter charts by dataSet

### LINKED ISSUE #15129: [native filter] native filter on different datasets can take effect on the dataset not in configuration when 'apply to all panel' is selected.
native filter on different data sets can take take effect on the dataset not in configuration when 'apply to all panel' is selected
For example, i have 2 charts in a dashboard:
time gain filter control chart from dataset: cleaned_sales_data

### How to reproduce the bug
1, create Time column filter test:
a. use 'FCC survey 2018' data set, edit dataset, select is temporal for 'time_end'
b, create time series chart using time column 'time end', metric 'MAX(month_job_search)User create time series chart using time column 'time end', metric 'MAX(month_job_search) add to dashboard 
c. create time column native filter in dashboard and set dataset to be 'FCC survey 2018', apply to all panel

2, create Time grain filter test:
a, use 'cleaned_sales_data' and create time-series chart on sum (sales)
b, create time grain native filter on 'cleaned_sales_data' dataset and apply to all panel

3, apply time column filter and time grain filter to observe change

### Expected results
 time column filter only take effect on the 'FCC survey 2018' time series chart and time grain filter only take effect on 'clean_sales_data' time series chart

### Actual results
time column filter and time grain filter take effect on both chart

#### Screenshots

Screen Shot 2021-06-10 at 8 53 07 PM

Screen Shot 2021-06-10 at 8 53 17 PM

https://user-images.githubusercontent.com/81597121/121756866-73d0bc80-cad0-11eb-8571-67e7bc8acfc3.mov

### Environment

superset master

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by simcha90**
### SUMMARY

Fixes: https://github.com/apache/superset/issues/15129

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/56388545/125189655-d806ad80-e241-11eb-9c34-6fb2d0259151.mov

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/dashboard/util/activeAllDashboardFilters.ts [hunk 1]
```diff
@@ -82,40 +90,65 @@ export const getRelevantDataMask = (
...
+  const chartDatasetId = charts?.[chartId]?.formData?.datasource?.split(
+    '__',
+  )?.[0];
+  const foundDatasetInNativeFilters =
+    nativeFilters[filterId] &&
+    (`${nativeFilters[filterId]?.targets?.[0]?.datasetId}` === chartDatasetId ||
+      !nativeFilters[filterId]?.targets?.[0]?.datasetId);
+  const foundDatasetInCrossFilters =
+    charts?.[filterId]?.formData?.datasource === chartDatasetId ||
+    !charts?.[filterId]?.formData?.datasource;
```
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/util/activeAllDashboardFilters.ts:108]:
```suggestion
  const datasetId = nativeFilters[filterId]?.targets?.[0]?.datasetId;
  const foundDatasetInNativeFilters =
    nativeFilters[filterId] &&
    (`${datasetId}` === chartDatasetId || !datasetId);

  const datasource = charts?.[filterId]?.formData?.datasource;
  const foundDatasetInCrossFilters =
    datasource === chartDatasetId || !datasource;
```

**Comment by michael-s-molina** [comment]
@simcha90 Can you change "Implements" to "Fixes" in the PR description? This way, when this PR is merged it will automatically close the issue.

**Comment by kgabryje** [comment]
@simcha90 The bug was reported to occur when "Apply to all panels" scope is selected. When you create a new filter, "Apply to specific panels" with charts from different datasets deselected is the default, which means that you run your tests with the charts that don't match deselected. However, this is an old issue. I would argue that it's no longer relevant, since we deselect charts that use different datasets by default. Selecting "Apply to all panels" is a conscious decision of the user. There are some cases when we want to apply a filter based on some dataset  to charts that use different dataset - that's why I didn't exclude those charts when using "Apply to all panels". See https://github.com/apache/superset/pull/15302.
CC @jinghua-qa @junlincc

**Comment by kgabryje** [comment]
Simcha, if Junlin and Jinghua confirm that the issue is not relevant anymore then I apologise for your wasted time. This issue should have been closed by #15302

**Comment by graceguo-supercat** [comment]
> Selecting "Apply to all panels" is a conscious decision of the user. There are some cases when we want to apply a filter based on some dataset to charts that use different dataset

Thank you @kgabryje. This "Apply to all panels" behavior **should not be changed**. This is critical for many dashboards in airbnb. If necessary, besides **"Apply to all panels"** and **"Apply to specific panel"**,  we could add another option say **"Apply to same dataset**.

**Review by graceguo-supercat — CHANGES_REQUESTED** [review]
This change will cause feature regression.

**Comment by simcha90** [comment]
@graceguo-supercat @kgabryje @junlincc but if I understood correctly there is still exists issue, consider next case:
1. Create charts A and B with different datasets 1 and 2
2. Create Filter that affects dataset 1 with selected chart A scope
3. Update chart A in explore to use dataset 2
So now we have case that Filter that still affects chart A but it already use dataset 2... same in opposite direction

**Comment by graceguo-supercat** [comment]
@simcha90 I can see the need to make filter only applicable to a dataset, which can fix the issue that your described. So i suggest to add a new option: **"Apply to all panels"** and "**Apply to specific panel**", and "**Apply to same dataset**". 
But should not remove the option "**Apply to all panels**".

**Pull request closed without merging**